### PR TITLE
dannysi/alert Alert tool (reopen, as the last one was closed by the force push) 

### PR DESCRIPTION
**Pull request by dannysi**

### CODE REVIEW

File: src/db_models.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ class Event(Base):
 class Timer(Base):
     __tablename__ = 'timers'
 
-    id = Column(Integer, primary_key=True)
+    id = Column(Integer, primary_key=True, server_default=text("nextval('timer_id_seq'::regclass)"))
```
Comment (yairl) [inline src/db_models.py:74]: Are you sure this is needed?

Searching says this happens automatically if primary_key=True.
I tried it as part of a different feature where this had to go, and seems to work fine.

Ideally - revert this change.
If not - let's talk about why this is needed.

File: src/services/messengers/messenger_factory.py [hunk 2]
```diff
@@ -1,6 +1,12 @@
-from typing import Dict
+from typing import Dict, Tuple
 from services.messengers.messenger import MessagingService
 from services.messengers.tg import TelegramMessenger
 from services.messengers.wa import WhatsappMessenger
 
 messenger_by_type = { 'tg' : TelegramMessenger(), 'wa' : WhatsappMessenger()} # type: Dict[str, MessagingService]
+
+def get_messenger_and_chat_id(messenger_chat_id:str) -> Tuple[MessagingService, str]:
```
Comment (yairl) [inline src/services/messengers/messenger_factory.py:8]: Created #57 to handle this in the future.

File: src/services/open_ai/query_openai.py [hunk 3]
```diff
@@ -164,12 +167,13 @@ def get_prep_message(ctx : Context, messenger, is_final : bool) -> Dict[str, str
 When providing a final answer, use this exact format, and do not deviate.
 IMPORTANT: ALWAYS wrap your final answer with <yair1xigoresponse> tags, and in JSON format.
 
-Today's date is {current_date}.
+Today's full date time is {current_date}.
```
Comment (yairl) [inline src/services/open_ai/query_openai.py:170]: no time -> drop the word

File: src/services/open_ai/query_openai.py [hunk 4]
```diff
@@ -164,12 +167,13 @@ def get_prep_message(ctx : Context, messenger, is_final : bool) -> Dict[str, str
...
 For up-to-date information about people, stocks and world events, ALWAYS use one of the tools available to you and DO NOT provide an answer.
 For fiction requests, use your knowledge and creativity to answer.
 If human request has no context of time, assume he is referring to current time period.
 All tools provided have real-time access to the internet; do not reply that you have no access to the internet, unless you have attempted to invoke the SEARCH tool first. Additionally, do not invoke a tool if the required TOOL_INPUT is unknown, vague, or not provided. Always follow the IMPORTANT note in the tool description.
 If you have missing data and ONLY if you cannot use the tools provided to fetch it, try to estimate; in these cases, let the user know your answer is an estimate.
+You are able to set reminders and when you are asked to do it you will invoke the ALERT tool.
```
Comment (yairl) [inline src/services/open_ai/query_openai.py:176]: 1. Was this needed for R1X to accept reminders?
2. This needs to clearly state reminder has to be in relative time.

File: src/services/open_ai/query_openai.py [hunk 5]
```diff
@@ -387,14 +397,20 @@ def invoke_tool(ctx:Context, tool, input):
...
     if tool_canon.startswith('WEATHER'):
         answer = invoke_weather_search(ctx, input)
 
-        return answer
+        return answer, False
+    
+    if tool_canon.startswith('ALERT'):
+        ctx.set_stat('tools-flow:tool-alert', 1)
+        invoke_alert_tool(ctx, input, message)
+        return "succesfully added the timer", True
```
Comment (yairl) [inline src/services/open_ai/query_openai.py:410]: Message: alert added successfully.

Best if you include the actual time it is set to.

File: src/services/timers.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+
+import datetime
+import time
+import traceback
+from typing import Tuple
+from infra import logger, utils 
+from infra.context import Context
+utils.load_env()
```
Comment (yairl) [inline src/services/timers.py:8]: We already load the environment at time 0 via src/run.py.
Why is it needed here?

File: tools/multi_sender.py [hunk 7]
```diff
@@ -6,18 +6,16 @@
 from typing import Dict, List
 
 sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'src')))
-from infra import utils
-from services.messengers.messenger import MessagingService
+from infra import utils 
+from infra.context import Context
```
Comment (yairl) [inline tools/multi_sender.py:10]: You are not using Context here, need to drop

File: src/services/timers.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+import db_models
+from services.messengers import messenger_factory
+
+def invoke_alert_tool(ctx:Context, alert_args:Tuple[int, str], parsed_message):
+    messenger_chat_id=f"{parsed_message.source}:{parsed_message.chatId}"
```
Comment (yairl) [inline src/services/timers.py:13]: spaces man...

messenger_chat_id = f"..."

instead of:

messenger_chat_id=f"fdsafdsa"

**Review by yairl — CHANGES_REQUESTED** [review]

File: src/services/open_ai/query_openai.py (hunk 5 above)
Comment (dannysi) [reply src/services/open_ai/query_openai.py:410]: Relative time? I don't think it's needed, if we had an absolute time, it would be cool

File: src/services/timers.py (hunk 6 above)
Comment (dannysi) [reply src/services/timers.py:8]: because it's a load of a different thread

File: tools/multi_sender.py (hunk 7 above)
Comment (dannysi) [reply tools/multi_sender.py:10]: line 16 has Context

File: src/services/timers.py (hunk 8 above)
Comment (dannysi) [reply src/services/timers.py:13]: You are right, of course, I have to set up my beautify plugins

File: src/services/open_ai/query_openai.py (hunk 4 above)
Comment (yairl) [reply src/services/open_ai/query_openai.py]: Need to check what happens in 3.5.

Put a couple of tests in test/data, then use local-test to check.

File: src/services/timers.py (hunk 6 above)
Comment (yairl) [reply src/services/timers.py:8]: Shouldn't matter, env variables are per processes.

Just make sure the first thing src/run.py does is load the env, and this isn't needed.

Comment (yairl) [inline src/services/timers.py:8]: This should happen in the main entry point, src/run.py, not here.
I will handle it in a follow-up commit :)

**Review by yairl — APPROVED** [review]
